crates/angstrom-net/src/manager.rs

```diff
@@ -119,13 +182,22 @@ impl<DB: Unpin> StromNetworkManager<DB> {
                 self.swarm.sessions_mut().send_message(&peer_id, msg)
             }
             StromNetworkHandleMsg::Shutdown(tx) => {
-                // Disconnect all active connections
-                self.swarm
+                let peers = self
+                    .swarm
                     .sessions_mut()
-                    .disconnect_all(Some(DisconnectReason::ClientQuitting));
+                    .active_sessions
+                    .iter()
+                    .map(|(_, session)| CachedPeer {
+                        peer_id: session.remote_id,
+                        addr:    session.socket_addr
+                    })
+                    .collect::<Vec<_>>();
 
-                // drop pending connections
+                Self::save_known_peers(&peers);
 
+                self.swarm
+                    .sessions_mut()
+                    .disconnect_all(Some(DisconnectReason::ClientQuitting));
                 let _ = tx.send(());
             }
```

Comment:
not sure if this is the actual path that gets hit, did you check to see if the shutdown message from the handle actually gets called in all scenarios.

Reply:
what ideally you can do is when we initialize our network. Do a spawn critical with shutdown and then when the task executor gives the shutdown signal, then save the peers. That way, we always will know it will write on shutdown